learning/courses/quantum-machine-learning/qvc-qnn.ipynb

```diff
@@ -0,0 +1,1988 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "38153b01-8832-4eca-9892-262465d0f137",
+   "metadata": {},
+   "source": [
+    "This is a reasonable two-qubit depth. We should be able to get high-quality results from a real quantum computer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "id": "10cb9de5-3893-4174-a36b-7607ea908721",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "11\n",
+      "2+ qubit depth: 5\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Combine the feature map and variational circuit\n",
+    "ansatz = qnn_circuit\n",
+    "\n",
+    "# Combine the feature map with the ansatz\n",
+    "full_circuit = QuantumCircuit(num_qubits)\n",
+    "full_circuit.compose(feature_map, range(num_qubits), inplace=True)\n",
+    "full_circuit.compose(ansatz, range(num_qubits), inplace=True)\n",
+    "\n",
+    "# Check the depth of the full circuit\n",
+    "print(full_circuit.decompose().depth())\n",
+    "print(\n",
+    "    f\"2+ qubit depth: {full_circuit.decompose().depth(lambda instr: len(instr.qubits) > 1)}\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "53c81996-7269-4aca-893a-53082287ef6f",
+   "metadata": {},
+   "source": [
+    "Because we are using the ```ZFeatureMap```, which has no CNOT gates, adding the encoding layer does not increase our two-qubit depth. We can visualize the full circuit here."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "id": "b43806aa-7d41-40b8-9e43-a604f0e42c26",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "<Image src=\"/learning/images/courses/quantum-machine-learning/qvc-qnn/extracted-outputs/b43806aa-7d41-40b8-9e43-a604f0e42c26-0.avif\" alt=\"Output of the previous code cell\" />"
+      ]
+     },
+     "execution_count": 69,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "full_circuit.decompose().draw(\"mpl\", style=\"clifford\", idle_wires=False, fold=-1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e9cbfbd8-337c-4f10-b2de-5de8165a8fd1",
+   "metadata": {},
+   "source": [
+    "You may note that if minimizing two-qubit depth were of paramount importance, we could actually reduce it a bit by changing the order of the CNOTs. For example, the CNOTs on $q_{35}$ and $q_{34}$ could be moved to the left in the circuit diagram above, and could be placed directly below the CNOTs on $q_{30}$ and $q_{31}$, for example. For a 2-qubit gate depth of 5, it isn't obvious that this will make a difference after transpilation, but it is something to keep in mind. If the order of the CNOT gates is important for logically matching the problem at hand, the depth here is fine. If the order of CNOTs is not critical to modeling the data structure in our images, then we could write a script to re-order these CNOT gates to minimize depth.\n",
```

Comment:
```suggestion
    "You may note that if minimizing two-qubit depth were of paramount importance, we could actually reduce it a bit by changing the order of the CNOTs. For example, the CNOTs on $q_{35}$ and $q_{34}$ could be moved to the left in the circuit diagram above, and could be placed directly below the CNOTs on $q_{30}$ and $q_{31}$, for example. For a two-qubit gate depth of 5, it isn't obvious that this will make a difference after transpilation, but it is something to keep in mind. If the order of the CNOT gates is important for logically matching the problem at hand, the depth here is fine. If the order of CNOTs is not critical to modeling the data structure in our images, then we could write a script to re-order these CNOT gates to minimize depth.\n",
```